schemachanger: defer catalog writes pre-commit

postamar opened the pull request:
Previously, in the declarative schema changer, each mutation stage would
be executed by visiting each of the stage's mutation ops to collect the
mutation state, followed by applying the mutations eagerly.

This commit defers the running of kv.Batch containing the catalog
mutations to the pre-commit phase. This change doesn't functionally
change anything but paves the way for undoing all pending catalog
mutations and re-planning the schema changes at pre-commit time.

Informs #88294.

Release note: None

Commit:
schemachanger: defer catalog writes pre-commit

Previously, in the declarative schema changer, each mutation stage would
be executed by visiting each of the stage's mutation ops to collect the
mutation state, followed by applying the mutations eagerly.

This commit defers the running of kv.Batch containing the catalog
mutations to the pre-commit phase. This change doesn't functionally
change anything but paves the way for undoing all pending catalog
mutations and re-planning the schema changes at pre-commit time.

Informs #88294.

Release note: None

Commit:
schemachanger: remove RemoveAllTableComments op

This operation is no longer required.

Release note: None

Commit:
sctestdeps: fix comments handling

This commit fixes a bug where comments weren't handled correctly by the
test state. This wasn't noticeable because the code wasn't actually
exercised anywhere. This commit also fixes this.

Release note: None

Commit:
datapathutils: fix RewritableDataPath

This commit fixes a bug with this function which only becomes apparent
when not using bazel. This only impacts testing with GoLand.

Release note: None

Commit:
parser: make COMMENT ON explainable

Previously, it was impossible to execute EXPLAIN (DDL) COMMENT ON...
This commit fixes this.

Release note (bug fix): EXPLAIN (DDL) COMMENT ON is now possible.

Commit:
rules: improve test stability

In some cases, we define a set of almost-identical rules with the same
name, purposefully. The name is what is used by the data-driven test
output to sort the rule definitions, because these otherwise appear in
the order that the init() functions were called, which is undefined.

Sorting by name reduced the instability of the data-driven test when
adding or removing init() blocks containing rule definitions, but did
not altogether remove it.

This commit should help further remove it by using a stable sort. In
this way, the test output should be stable assuming that rule names are
not re-used across init() blocks.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/95330)
<!-- Reviewable:end -->

Commit:
schemachangerccl: rename generated tests

This commit adds an underscore in the generated tests' name where there
previously was one missing.

Release note: None

ajwerner commented on pkg/sql/opt/exec/execbuilder/testdata/show_trace:
```diff
@@ -167,10 +167,6 @@ $trace_query
 sql query       rows affected: 0
 commit sql txn  CPut /Table/3/1/109/2/1 -> table:<name:"kv2" id:109 version:3 modification_time:<> parent_id:106 unexposed_parent_schema_id:107 columns:<name:"k" id:1 type:<family: IntFamily width: 64 precision: 0 locale: "" visible_type: 0 oid: 20 time_precision_is_set: false > nullable:true hidden:false inaccessible:false generated_as_identity_type:NOT_IDENTITY_COLUMN virtual:false pg_attribute ...
 commit sql txn  Del /NamespaceTable/30/1/106/107/"kv2"/4/1
-commit sql txn  DelRange /Table/24/1/1/109
-commit sql txn  DelRange /Table/24/1/2/109
-commit sql txn  DelRange /Table/24/1/3/109
-commit sql txn  DelRange /Table/24/1/5/109
```
Do you have a clear understanding as to why this went away?

postamar replied:
Yes, `9b8eb53881ca41d41335271af68d6b9461211bb0` removes the `RemoveAllTableComments` op which was completely redundant and which generated those `DelRange`s.

ajwerner commented on pkg/ccl/schemachangerccl/BUILD.bazel:
```diff
@@ -51,7 +51,7 @@ sctest_gen(
     new_cluster_func = "newCluster",
     package = "schemachangerccl",
-    suffix = "base",
+    suffix = "_base",
```
Thank you

ajwerner commented on pkg/sql/schemachanger/testdata/explain_verbose/drop_table:
```diff
@@ -392,14 +399,29 @@ EXPLAIN (ddl, verbose) DROP TABLE db.sc.t;
+│           ├── • RemoveTableComment
 │           │     TableID: 107
 │           │
+│           ├── • LogEvent
+│           │     Element:
+│           │       TableComment:
+│           │         comment: t has a comment
+│           │         tableId: 107
+│           │     EventBase:
+│           │       Authorization:
+│           │         UserName: root
+│           │       Statement: DROP TABLE ‹db›.‹sc›.‹t›
+│           │       StatementTag: DROP TABLE
+│           │       TargetMetadata:
+│           │         SourceElementID: 1
+│           │         SubWorkID: 1
+│           │     TargetStatus: 1
```
I can't decide whether this is a good thing to show. It's fine for now but we might end up generating a lot of events when dropping tables.

postamar replied:
I'm working on this. The event logs are a bit of a nightmare to work with in the declarative schema changer. We should just roll out a new kind of event message.

ajwerner commented on pkg/sql/schemachanger/testdata/explain_verbose/drop_column_basic.rollback_1_of_7:
```diff
@@ -72,13 +80,36 @@ EXPLAIN (ddl, verbose) rollback at post-commit stage 1 of 7;
             │     Name: j
             │     TableID: 104
             │
+            ├── • UpsertColumnComment
+            │     ColumnID: 2
+            │     Comment: j has a comment
+            │     PGAttributeNum: 2
+            │     TableID: 104
+            │
+            ├── • LogEvent
+            │     Element:
+            │       ColumnComment:
+            │         columnId: 2
+            │         comment: j has a comment
+            │         pgAttributeNum: 2
+            │         tableId: 104
+            │     EventBase:
+            │       Authorization:
+            │         UserName: root
+            │       Statement: ALTER TABLE ‹defaultdb›.public.‹t› DROP COLUMN ‹j›
+            │       StatementTag: ALTER TABLE
+            │       TargetMetadata:
+            │         SourceElementID: 1
+            │         SubWorkID: 1
+            │     TargetStatus: 2
```
I wonder about this. Perhaps we want to not remove the comment until we hit the non-revertible stage.

postamar replied:
I don't know, it seems like a side-effect we'd like to see in-txn.

postamar replied:
I'm working towards that, but this test output changed only because I added a comment, not because I changed any behaviour, so eager comment removal is already a thing right now.

ajwerner commented on pkg/testutils/datapathutils/data_path.go:
```diff
@@ -64,7 +65,18 @@ func RewritableDataPath(t testing.TB, relative ...string) string {
 		t.Fatal("unable to get caller information")
 	}
-	cockroachWorkspace := filepath.Dir(filepath.Dir(filepath.Dir(thisFilePath)))
+	cockroachWorkspace := thisFilePath
```
what is this about? isn't `testutils/datapathutils/data_path.go` always at a fixed depth from the workspace dir?

ajwerner reviewed (commented):
This could have been a few commits or PRs. I think the comment changes are welcome but maybe need some attention regarding their revertibility. The testutils change I don't get. In general, it looks good.

postamar commented:
Closed in favour of https://github.com/cockroachdb/cockroach/pull/95467 which just has small fixes.
